### PR TITLE
Correct khr Extension Checks for Device Versions

### PR DESCRIPTION
**Pull request by FranklandJack**
* Correct khronos extension device version checks so that 2.0, 2.1 and
2.2 extensions are only require for devices versions < 3.0
(`cl_khr_3d_image_writes`, `cl_khr_image2d_from_buffer` and
`cl_khr_depth_images` are all optional for devices supporting
OpenCL-3.0).

* Remove `cl_khr_fp64` from list or required extensions for 2.0,
`cl_khr_fp64` is a required extension for OpenCL-1.2 devices if they
suport doubles. Check for support of `cl_khr_fp64` on 1.2 devices.

* Remove duplicated `cl_khr_byte_addressable_store` from list of 2.0
extensions. Since this is a 1.1 extension and all versions greater than
1.1 are backwards compatible with 1.1 this will be checked anyway.

### CODE REVIEW

File: test_conformance/api/test_api_min_max.cpp [hunk 1]
```diff
@@ -1946,64 +1946,90 @@ int test_min_max_device_version(cl_device_id deviceID, cl_context context, cl_co
         return -1;
     }
```
Comment (kpet) [inline test_conformance/api/test_api_min_max.cpp:1948]: Please replace all the version parsing code above with `get_device_cl_version`.

File: test_conformance/api/test_api_min_max.cpp [hunk 2]
```diff
@@ -1946,64 +1946,90 @@ int test_min_max_device_version(cl_device_id deviceID, cl_context context, cl_co
...
 
-    if( major * 10 + minor >= 11 )
+
+    int version_as_integer = major * 10 + minor;
+    if (version_as_integer >= 11)
     {
         log_info( "Checking for required extensions for OpenCL 1.1 and later devices...\n" );
-        for( i = 0; NULL != requiredExtensions[i]; i++ )
+        for (i = 0;
+             i < sizeof(requiredExtensions11) / sizeof(requiredExtensions11[0]);
```
Comment (kpet) [inline test_conformance/api/test_api_min_max.cpp:1973]: Please use `ARRAY_SIZE`

File: test_conformance/api/test_api_min_max.cpp [hunk 3]
```diff
@@ -1946,64 +1946,90 @@ int test_min_max_device_version(cl_device_id deviceID, cl_context context, cl_co
...
-            // if double precision is not supported, then don't check extension "cl_khr_fp64"
-            numRequiredExtension20 -= (value == 0);
-
-            for( i = 0; i<numRequiredExtension20; i++ )
+        if (version_as_integer >= 20 && version_as_integer < 30)
+        {
+            log_info("Checking for required extensions for OpenCL 2.0, 2.1 and "
+                     "2.2 devices...\n");
+            for (i = 0; i < sizeof(requiredExtensions2x)
+                     / sizeof(requiredExtensions2x[0]);
```
Comment (kpet) [inline test_conformance/api/test_api_min_max.cpp:2019]: `ARRAY_SIZE`

File: test_conformance/api/test_api_min_max.cpp (hunk 2 above)
Comment (FranklandJack) [reply test_conformance/api/test_api_min_max.cpp]: Thanks for your feedback, I have updated to use `ARRAY_SIZE`.

File: test_conformance/api/test_api_min_max.cpp (hunk 3 above)
Comment (FranklandJack) [reply test_conformance/api/test_api_min_max.cpp]: Thanks for your feedback, I have updated to use `ARRAY_SIZE`.

File: test_conformance/api/test_api_min_max.cpp (hunk 1 above)
Comment (FranklandJack) [reply test_conformance/api/test_api_min_max.cpp:1898]: Thanks for your feedback. I have update te parsing code, I wasn't sure about the checking of the `OPENCL_1_0_DEVICE` and `OPENCL_1_1_DEVICE`, I assume we still want to keep these checks so I've left them in.

Comment (kpet) [reply test_conformance/api/test_api_min_max.cpp:1898]: I'd get rid of these, I think they were used to adapt the CTS behaviour to 1.1 and 1.0 implementations. We're now using queries for this.

Comment (FranklandJack) [reply test_conformance/api/test_api_min_max.cpp]: Thanks, okay that makes sense, I have removed them and reduced the scope of the loop counters and the `error` variable.

**Pull request merged**